src/jsonrpc.rs

```diff
@@ -156,26 +190,69 @@ fn process_http_request(obj: JsonRequest, infura_url: &str) -> Vec<u8> {
     }
 }
 
-fn handle_request(obj: JsonRequest, infura_url: &str) -> Result<String, String> {
-    let request_id = obj.id;
-    match obj.method.as_str() {
-        "web3_clientVersion" => Ok(format!(
-            r#"{{"jsonrpc":"2.0","id":{},"result":"trin 0.0.1-alpha"}}"#,
-            request_id,
+fn handle_request(
+    obj: JsonRequest,
+    infura_url: &str,
+    portal_tx: Sender<PortalEndpoint>,
+) -> Result<String, String> {
+    let method = obj.method.as_str();
+    if method.starts_with("discv5") {
+        dispatch_portal_request(obj, portal_tx)
+    } else {
+        match method {
+            "web3_clientVersion" => Ok(format!(
+                r#"{{"jsonrpc":"2.0","id":{},"result":"trin 0.0.1-alpha"}}"#,
+                obj.id,
+            )),
+            _ => dispatch_infura_request(obj, infura_url),
+        }
+    }
+}
+
+fn dispatch_infura_request(obj: JsonRequest, infura_url: &str) -> Result<String, String> {
+    //Re-encode json to proxy to Infura
+    let request = serde_json::to_string(&obj).unwrap();
+    match proxy_to_url(request, infura_url) {
+        Ok(result_body) => Ok(std::str::from_utf8(&result_body).unwrap().to_owned()),
+        Err(err) => Err(format!(
+            r#"{{"jsonrpc":"2.0","id":"{}","error":"Infura failure: {}"}}"#,
+            obj.id,
+            err.to_string(),
         )),
+    }
+}
+
+#[tokio::main]
+async fn dispatch_portal_request(
```

Comment:
You can put `tokio::main` at any boundary that starts an async context, but yeah, I think it would probably be better to hoist this as far up as possible.

Reply:
> ... but yeah, I think it would probably be better to hoist this as far up as possible.

Just to confirm my understanding, you're suggesting making all the functions that are called b/w `main()` and here `async`? Any suggestion on which approach is better b/w this or using an unbounded channel?

Reply:
To quote from the [tokio docs](https://docs.rs/tokio/1.8.1/tokio/attr.main.html):

> [...] starting a new runtime each time it is called. If the function is called often, it is preferable to create the runtime using the runtime builder so the runtime can be reused across calls.

Using the `tokio::main` attribute fires up all the machinery for tokio every time the annotated function is called, so I think we probably should move it up to the top `fn main()`. On the other hand, we shouldn't just convert every non-async function all the way up to the entry point into an async one, especially if they contain I/O calls.

I think, for now, the best approach would be to use an `UnboundedSender` to bridge the sync/async worlds.

Thoughts @carver?

Reply:
Sweet! Thanks for the clarification - I implemented the `UnboundedSender` and got rid of the excessive `tokio::main`. I'd always welcome a final :eyes: from you or @carver to double check, but everything is functioning as expected locally so i'll go ahead with the merge unless i hear otherwise by eod.